ISSUE-114 Better handling of Generic/Specific Records

michaelandrepearce opened the pull request:
Allow producer with SpecificRecord and consumer with GenericRecord and vice versa.

This deprecates the use of the recordType in the protocol but we leave reading it, to keep compatbility.

satishd commented on schema-registry/serdes/src/main/java/com/hortonworks/registries/schemaregistry/serdes/avro/AbstractAvroSnapshotDeserializer.java:
```diff
@@ -138,19 +151,13 @@ protected Object buildDeserializedObject(InputStream payloadInputStream,
                 // generate UTF-8 string object from the received bytes.
                 deserializedObj = new String(IOUtils.toByteArray(payloadInputStream), AvroUtils.UTF_8);
             } else {
+                //This is deprecated but kept to keep protocol.
```
I am fine with removing this in serializer also breaking compatibility with earlier version. It is better to avoid this compatibility baggage in these early initial versions. We will maintain backward compatibility from the upcoming release.

michaelandrepearce replied:
ok sure, ill remove from producer also.

satishd commented on schema-registry/serdes/src/main/java/com/hortonworks/registries/schemaregistry/serdes/avro/AbstractAvroSnapshotDeserializer.java:
```diff
@@ -159,4 +166,39 @@ protected Object buildDeserializedObject(InputStream payloadInputStream,
+        if(readerSchema == null) {
+            Class readerClass = SpecificData.get().getClass(writerSchema);
+            if(readerClass == null) {
+                throw new SerializationException("Could not find class " + writerSchema.getFullName() + " specified in writer\'s schema whilst finding reader\'s schema for a SpecificRecord.");
```
We may want to use `SerDesException` here. Better not to use kafka specific `SerializationException` as this class is for avro, not specifically for kafka.

michaelandrepearce replied:
sure will do :)

satishd commented on schema-registry/serdes/src/main/java/com/hortonworks/registries/schemaregistry/serdes/avro/AbstractAvroSnapshotDeserializer.java:
```diff
@@ -159,4 +166,39 @@ protected Object buildDeserializedObject(InputStream payloadInputStream,
+            try {
+                readerSchema = ((SpecificRecord)readerClass.newInstance()).getSchema();
+            } catch (InstantiationException var5) {
+                throw new SerializationException(writerSchema.getFullName() + " specified by the " + "writers schema could not be instantiated to find the readers schema.");
```
same as earlier comment.

michaelandrepearce replied:
ditto :)

satishd commented on schema-registry/serdes/src/main/java/com/hortonworks/registries/schemaregistry/serdes/avro/AbstractAvroSnapshotDeserializer.java:
```diff
@@ -159,4 +166,39 @@ protected Object buildDeserializedObject(InputStream payloadInputStream,
+            } catch (InstantiationException var5) {
+                throw new SerializationException(writerSchema.getFullName() + " specified by the " + "writers schema could not be instantiated to find the readers schema.");
+            } catch (IllegalAccessException var6) {
+                throw new SerializationException(writerSchema.getFullName() + " specified by the " + "writers schema is not allowed to be instantiated to find the readers schema.");
```
same as earlier comment.

michaelandrepearce replied:
hahaha :) okies :)

satishd commented:
@michaelandrepearce You may want to add few tests covering some of the cases this change is done for.

michaelandrepearce commented:
@satishd added tests, removed deprecated type in the protocol (aka serialiser no longer adds, deserialiser no longer reads/jumps over it) 


To add tests i had to allow myself to inject a mocked schema registry client, so we could make simple but really useful unit tests.

Whilst I added what we needed to.

I noted that the interfaces ISchemaRegistryClient and ISchemaRegistry whilst almost equal, very annoyingly differ almost everywhere, some places just a slight method name difference. 

If you are having a breaking release could we strongly suggest making one interface which then both server side implementation and client side implement, this would make life a lot easier. e.g. theres methods named slightly differently, there methods with different return objects.

harshach commented:
@michaelandrepearce Thanks for your interest and contributions to the registry project. I emailed you ICLA doc . Can you please sign that and send it to me. Thanks.

michaelandrepearce commented:
Which email? I don't see anything in either my private or work emails :(

satishd commented:
>If you are having a breaking release could we strongly suggest making one interface which then both server side implementation and client side implement, this would make life a lot easier. e.g. theres methods named slightly differently, there methods with different return objects.

@michaelandrepearce Both of them have been slightly diverged, we can address without breaking any backward compatibility of `SchemaRegistryClient`.

Currently busy with an internal release, I will review and put comments on other PRs in next 2 days.

michaelandrepearce commented:
@satishd it would be good to do that :) , as seen this doesn't block this work, as managed to make a mapping for now for all the methods needed for a mock client. 
Would you propose to make bring inline the ISchemaRegistry then so the client isn't affected? Maybe we should move this to a new backlog issue?

satishd commented:
>Would you propose to make bring inline the ISchemaRegistry then so the client isn't affected? 

Right.

>Maybe we should move this to a new backlog issue?

Sure, we should address this in a separate issue.

michaelandrepearce commented:
raised https://github.com/hortonworks/registry/issues/130

satishd commented:
Closes #114

michaelandrepearce commented:
@satishd i doesn't look like this got merged as once you merged the other PR there were conflicts. I have rebased and squashed which has resolved. 

good to merge?

satishd commented:
@michaelandrepearce I resolved the conflicts and found few tests are failed which I am fixing before the merge. Anyways, I will take your updated commit, fix the failing tests and merge the changes.

michaelandrepearce commented:
ah sorry, :(, but thanks, is there a CI (Jenkins or something) plan, linked with the repo?

satishd commented:
@michaelandrepearce no worries. :) We will have CI integration soon for public repo.
